staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -221,7 +223,11 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			// rule is empty
 			continue
 		}
-		if compiled.TransitionRule && oldObj == nil {
+
+		// If ratcheting is enabled, allow rule with oldSelf to evaluate
+		// when `optionalOldSelf` is set to true
+		allowsNilOldSelf := utilfeature.DefaultFeatureGate.Enabled(features.CRDValidationRatcheting) && rule.OptionalOldSelf != nil && !*rule.OptionalOldSelf
+		if compiled.UsesOldSelf && oldObj == nil && !allowsNilOldSelf {
```

Comment:
And you automatically get `oldSelf` as being nil in the following code?

Reply:
No, had to update the activation factory to also return `true` for `hasOldSelf`